File: algorithms/python/printWordsVertically/printWordsVertically.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/print-words-vertically/
# Author: Miao Zhang
# Date:   2021-04-23

class Solution:
    def printVertically(self, s: str) -> List[str]:
        words = s.split()
        maxlen = max(len(word) for word in words) # row
        res = list()
        for i in range(maxlen):
            ss = ''
            for word in words:
                if i < len(word):
                    ss += word[i]
                else:
                    ss += ' '
            res.append(ss.rstrip())
        return res

File: algorithms/cpp/printWordsVertically/printWordsVertically.cpp
// Source: https://leetcode.com/problems/print-words-vertically/
// Author: Miao Zhang
// Date:   2021-04-23

class Solution {
public:
    vector<string> printVertically(string s) {
        stringstream ss(s);
        string word;
        vector<string> words;
        int maxlen = 0;
        while (ss >> word) {
            words.push_back(word);
            maxlen = max(maxlen, (int)word.size());
        }
        
        vector<string> res;
        // row
        for (int i = 0; i < maxlen; i++) {
            string tmp;
            for (string& word: words) {
                tmp += (i < word.size() ? word[i] : ' ');
            }
            while (tmp.back() == ' ') tmp.pop_back();
            res.push_back(tmp);
        }
        return res;
    }
};
